### DISCUSSION TITLE
Linux installation steps

### DISCUSSION THREAD

**Discussion by glopesdev** [bonsai-rx/bonsai#1101] [Articles]
Installing Bonsai on Linux
---------------------------

1. Install the [Mono](https://www.mono-project.com/download/stable/#download-lin) framework for Linux.
2. Download the portable release of [Bonsai 2.8 (.zip)](https://github.com/bonsai-rx/bonsai/releases/download/2.8.0/Bonsai.zip).
3. Extract the contents of the archive into a folder and run the following command `mono Bonsai.exe`.
4. Install **Bonsai - Starter Pack** from the package manager.
5. Go to the **Updates** tab in the package manager, select the **Include prerelease** checkbox and install all updates.

There is a common warning message related to libcanberra-gtk that can be removed by installing the relevant package:
```
sudo apt-get install libcanberra-gtk-module
```

## Install OpenGL

Several Bonsai packages require OpenGL 4.0+ for interfacing with graphics. Below are some suggested instructions for a vanilla Ubuntu distribution. They may have to be tweaked for your particular system:

```
sudo apt install mesa-utils
glxinfo | grep "OpenGL version"
sudo apt install libsdl2-dev
```

The `glxinfo` command should return version 4.0 or greater for full compatibility with the shaders package.

## Install OpenAL

The audio capture requires OpenAL as a dependency for interfacing with the microphone drivers.

```
sudo apt-get install libopenal1 libopenal-dev
sudo apt-get install libalut0 libalut-dev
```

## Configure access to serial port

Occasionally the access to serial port on devices such as an Arduino might be blocked by your system. If this is the case, you can often use `screen` to configure access to the serial port:

```
screen /dev/ttyUSB0
```

## Install OpenCV from pre-built binaries

Bonsai requires the C interface exposed by OpenCV 2.4 to drive its image processing primitives. OpenCV.NET provides pre-built binaries for a few ubuntu architectures on its [latest release 3.4.2](https://github.com/horizongir/opencv.net/releases/tag/v3.4.2).

### Add the earlier release repositories for libjasper

```
sudo add-apt-repository "deb http://security.ubuntu.com/ubuntu xenial-security main"
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 3B4FE6ACC0B21F32
sudo apt update
```

### Download and install pre-built binaries

The OpenCV project does not provide binary distributions for 2.4 for the latest versions of Ubuntu. These distributions are unofficial builds so you will have to download the .deb packages and install them locally.

```
wget https://github.com/horizongir/opencv.net/releases/download/v3.4.2/OpenCV-2.4.13.6-ubuntu-x86_64-libs.deb
wget https://github.com/horizongir/opencv.net/releases/download/v3.4.2/OpenCV-2.4.13.6-ubuntu-x86_64-dev.deb
sudo apt install OpenCV-2.4.13.6-ubuntu-x86_64-libs.deb
sudo apt install OpenCV-2.4.13.6-ubuntu-x86_64-dev.deb
```

## Build OpenCV from source

You can also build OpenCV from source for your system. The below guide is adapted from [the official build instructions](https://docs.opencv.org/2.4/doc/tutorials/introduction/linux_install/linux_install.html).

### Add the earlier release repositories for libjasper
```
sudo add-apt-repository "deb http://security.ubuntu.com/ubuntu xenial-security main"
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 3B4FE6ACC0B21F32
sudo apt update
```

### Install compiler dependencies

```
sudo apt-get install build-essential
sudo apt-get install cmake git libgtk2.0-dev pkg-config libavcodec-dev libavformat-dev libswscale-dev
sudo apt-get install python-dev python-numpy libtbb2 libtbb-dev libjpeg-dev libpng-dev libtiff-dev libjasper-dev libdc1394-22-dev
```

### Build and install OpenCV 2.4

Make sure to build the 2.4 branch, or [download the tarball from GitHub](https://github.com/opencv/opencv/archive/refs/tags/2.4.13.7.tar.gz).

For example to clone the 2.4 branch:
```
git clone -b 2.4 https://github.com/opencv/opencv.git
```

Example build configuration using cmake:
```
cd ~/opencv
mkdir release
cd release
cmake -D CMAKE_BUILD_TYPE=RELEASE -D CMAKE_INSTALL_PREFIX=/usr/local ..
```

Build using make:
```
make
sudo make install
```

### Configure environment

OpenCV installs libraries into `usr/local/lib`, so you may need to configure your environment using [one of these options](https://unix.stackexchange.com/questions/67781/use-shared-libraries-in-usr-local-lib) to share libraries at a system level.

**Comment by jbscherrer** [comment]
Is there a reason to expect Bonsai on linux to have better timing characteristics than on Windows? I'm trying to get feedback latency down as low as possible and wondering if switching to linux might be helpful.

**Reply by glopesdev** [reply]
With Mono I doubt you will see any improvements since Microsoft's implementation of generational GC was always superior. With the new .NET core however this picture might have changed, but we only have that update planned for next year's roadmap.

This Mono version was intended mostly as a preview and teaching aide but it does have the advantage of allowing to run Bonsai embedded on Raspberry Pis which might be useful for some simple embedded applications.

**Comment by roaldarbol** [comment]
Amazing work on Bonsai 2.8🥳 Can’t wait to run it on the Raspberry! Do you have a recommended way to install OpenCV for RPi? (I assume it’s slightly different from Ubuntu - or would you run Ubuntu instead of RPi OS?)

**Comment by ncguilbeault** [comment]
## Updated - Building and installing OpenCV from source on Ubuntu 22.04

When trying to build OpenCV from source on Ubuntu 22.04, you may encounter some errors during the build process. I was able to successfully build and utilize OpenCV version 2.4.1 for Bonsai-Rx on Ubuntu 22.04 using the following updated steps.

Add old libjasper repository to package list:

```
sudo add-apt-repository "deb http://security.ubuntu.com/ubuntu xenial-security main"
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 3B4FE6ACC0B21F32
sudo apt-key export C0B21F32 | sudo gpg --dearmour -o /etc/apt/trusted.gpg.d/libjasper.gpg
sudo apt update
```
Install compiler dependencies:
```
sudo apt-get install build-essential curl
sudo apt-get install cmake git libgtk2.0-dev pkg-config libavcodec-dev libavformat-dev libswscale-dev
sudo apt-get install libtbb2 libtbb-dev libjpeg-dev libpng-dev libtiff-dev libjasper-dev
```
OpenCV requires python 2, which no longer ships natively with Ubuntu. To install python 2 and numpy, you can run the following commands:
```
sudo apt-get install python2-dev
```
You then need to install the pip package manager for python 2, as well as the numpy library for python 2 using the following commands:
```
curl https://bootstrap.pypa.io/pip/2.7/get-pip.py --output get-pip.py
sudo python2 get-pip.py
sudo rm get-pip.py
python2 -m pip install numpy
```
You can install libdc with the following command:
```
sudo apt-get install libdc1394-dev
```
You can now clone the opencv repo:
```
git clone -b 2.4 https://github.com/opencv/opencv.git
```
And prepare to install using:
```
cd ~/opencv
mkdir release
cd release
```
This time, when running the make command, specify the paths to the python2 executable, include directory, and libraries. Specifying the c++ standard is also required:
```
cmake -D CMAKE_BUILD_TYPE=RELEASE -D CMAKE_INSTALL_PREFIX=/usr/local -D PYTHON_EXECUTABLE=/usr/bin/python2 -D PYTHON_INCLUDE_DIR=/usr/include/python2.7 -D PYTHON_LIBRARY=/usr/lib/x86_64-linux-gnu/libpython2.7.so -D CMAKE_CXX_STANDARD=11 ..
```
Finally, run to make the install script and install the OpenCV library:
```
make -j$(nproc)
sudo make install
```
For bonsai to access the OpenCV libraries, you can create an empty file to dynamically link the OpenCV library at the global system level:
```
sudo nano /etc/ld.so.conf.d/libopencv.conf
```
and add the following line to this file:
```
/usr/local/lib
```
Finally, run this command in the terminal to update:
```
sudo ldconfig
```

**Reply by chiyu1203** [reply]
Hi! @ncguilbeault  I followed your post to install Bonsai on Ubuntu 22.04. The installation looked fine as I was able to go through the entire procedure. However, when I ran Bonsai and use the CameraCapture module, I bumped into this error when using OpenCV library. Have you seen error this before? 
HIGHGUI ERROR: V4L/V4L2: VIDIOC_S_CROP
HIGHGUI ERROR: V4L2: Pixel format of incoming image is unsupported by OpenCV

**Reply by chiyu1203** [reply]
I uninstalled the old packages and recompiled everything with python3-based OpenCV based on @glopesdev original instruction (Build OpenCV from source) and solved the issue.

**Reply by ncguilbeault** [reply]
Hi @chiyu1203, glad you got it working. I have seen this error before as well. Installing the Video4Linux libraries solved this issue for me like so:
```
sudo apt update
sudo apt install v4l-utils libv4l-dev
```

**Reply by chiyu1203** [reply]
@ncguilbeault, thanks for putting bonsai-docker together! I have not tried that but it looks promising!

**Comment by caniko** [comment]
This is a great start for Linux support, but I would appreciate a dedicated page that is versioned and goes beyond Debian. I suggest we include guides for Fedora and Arch. I would love to have nix support, but understand that it is well beyond the scope of the project. Can we make this possible @glopesdev?

**Opinion**
I prefer mostly NixOS and Arch to a lesser extent over Ubuntu. I have created a request to support Bonsai on [nix](https://github.com/NixOS/nixpkgs/issues/338571), please vote up if you are using nix.

**Comment by banchan86** [comment]
Since my laptop is a mac, and I have been working with documentation using docfx on mac, i decided for fun to try running bonsai with mono (V2.8.5, v2.9.0-preview1) on Mac and unfortunately it doesn't work because it looks for Nuget.Config in the wrong place:

`NuGet.Configuration.NuGetConfigurationException: Failed to read NuGet.Config due to unauthorized access. Path: '/Users/XX/.config/NuGet/NuGet.Config'. `

--→ I found it instead in `/Users/XX/.nuget/NuGet/NuGet.Config`

No real reason to fix it since I don't know who would use it outside of my very specific workcase haha and I am sure other stuff will break down the road, just wanted to put this somewhere in case anyone was wondering and as a followup to the last time this question was posted on the google groups discussion https://groups.google.com/g/bonsai-users/c/5E6KQPzvsRs. Haven't tried VMs.

**Reply by banchan86** [reply]
Got bonsai on mac running with @ncguilbeault bonsai dockers https://github.com/ncguilbeault/bonsai-docker/tree/dev!

**Reply by chiyu1203** [reply]
@banchan86 That's amazing! just curious, did you manage to run the bonsai-base image or bonsai-vision? If it was latter, did you install opencv2.4 via Homebrew or Macports?

**Reply by banchan86** [reply]
I only ran the bonsai-base image so don't have any experience with the bonsai-vision image unfortunately

**Comment by glopesdev** [comment]
On a random tangent we got BonVision running in Bonsai 2.8.0 on a Raspberry Pi 4 model B by modifying shader code to compile to OpenGL ES 3.0.

Only gratings, circle and checkerboard primitives supported for now, images should work if native OpenCV 2.4.x build is available.

Zip files below, will edit later with more details:
- [BonVision.zip](https://github.com/user-attachments/files/18132606/BonVision.zip)
- [Bonsai.zip](https://github.com/user-attachments/files/18132617/Bonsai.zip)

**Comment by Mtkostecki** [comment]
Hi! 
I have managed to make Bonsai more or less usable, but I face two problems. When I use camera capture, I can do it only once - when i stop the program and try to run it again, it gives an error - failed to capture a frame. After restarting Bonsai it works again.

The second problem is that I cannot crop anything, or select an ROI - clocking on a square in properties has no effect, or sometimes even it makes Bonsai to stop working at all!

**Reply by ncguilbeault** [reply]
Glad to hear you got it working! 

It was a while ago but I do remember having the same issue you described with the camera not being reset properly. Just curious, did you try installing the `v4l-utils` and `libv4l-dev` libraries and see if this helps? 

As for the crop node, we've noticed this on our end as well. I believe it is related to a bug with mono. Until this mono bug gets fixed or we move to a better cross-platform solution, you can still set the `RegionOfInterest` property by directly editing the text. It's not ideal, but it should work if you manually type out the values you want in the format `X, Y, Width, Height`.

**Comment by skyshi0** [comment]
Hi, I encountered some issues running 'spinnaker' package in Linux. The package says that we should use Spinnaker.SDK 1.29.0.5 version, however, I only found the download links for x64_bionic(Ubuntu 18.04) and x64_xenial(Ubuntu 16.04). I'm currently using Spinnaker 3.1.0.79  and I encountered some errors using the SpinnakerCapture node, does anyone encounter similar issues for that? 

Method '<Module>:_getFiberPtrId ()' in assembly '/home/skyshi/Desktop/Software/Bonsai/Packages/Bonsai.Spinnaker.0.7.1/build/net462/bin/x86/SpinnakerNET_v140.dll' contains native code that cannot be executed by Mono on this platform. The assembly was probably created using C++/CLI.System.TypeInitializationException: The type initializer for '<Module>' threw an exception. ---> <CrtImplementationDetails>.ModuleLoadException: The C++ module failed to load. ---> System.MissingMethodException: Method contains unsupported native code assembly:<unknown assembly> type:<unknown type> member:(null)  at (wrapper managed-to-native) <Module>._getFiberPtrId()  at <Module>.<CrtImplementationDetails>.LanguageSupport._Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x0001c] in <0541963a254849e89638f912809c8605>:0  at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00028] in <0541963a254849e89638f912809c8605>:0   --- End of inner exception stack trace ---  at <Module>.<CrtImplementationDetails>.ThrowModuleLoadException (System.String errorMessage, System.Exception innerException) [0x00007] in <0541963a254849e89638f912809c8605>:0  at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00041] in <0541963a254849e89638f912809c8605>:0  at <Module>..cctor () [0x00008] in <0541963a254849e89638f912809c8605>:0   --- End of inner exception stack trace ---  at Bonsai.Spinnaker.SerialNumberConverter.GetStandardValues (System.ComponentModel.ITypeDescriptorContext context) [0x00006] in <120112d5098e4d53a7e9be3f2b34edfc>:0  at System.Windows.Forms.PropertyGridInternal.GridEntry.get_AcceptedValues () [0x00023] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.PropertyGridInternal.PropertyGridView.UpdateItem (System.Windows.Forms.PropertyGridInternal.GridEntry entry) [0x000b9] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.PropertyGridInternal.PropertyGridView.SelectItem (System.Windows.Forms.PropertyGridInternal.GridEntry oldItem, System.Windows.Forms.PropertyGridInternal.GridEntry newItem) [0x0000d] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at (wrapper remoting-invoke-with-check) System.Windows.Forms.PropertyGridInternal.PropertyGridView.SelectItem(System.Windows.Forms.PropertyGridInternal.GridEntry,System.Windows.Forms.PropertyGridInternal.GridEntry)  at System.Windows.Forms.PropertyGrid.SelectItemCore (System.Windows.Forms.PropertyGridInternal.GridEntry oldItem, System.Windows.Forms.PropertyGridInternal.GridEntry item) [0x0000e] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.PropertyGrid.set_SelectedGridItem (System.Windows.Forms.GridItem value) [0x0001e] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at (wrapper remoting-invoke-with-check) System.Windows.Forms.PropertyGrid.set_SelectedGridItem(System.Windows.Forms.GridItem)  at System.Windows.Forms.PropertyGridInternal.PropertyGridView.set_SelectedGridItem (System.Windows.Forms.PropertyGridInternal.GridEntry value) [0x00000] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.PropertyGridInternal.PropertyGridView.OnMouseDown (System.Windows.Forms.MouseEventArgs e) [0x000ad] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.Control.WmLButtonDown (System.Windows.Forms.Message& m) [0x0008b] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.Control.WndProc (System.Windows.Forms.Message& m) [0x001bc] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.ScrollableControl.WndProc (System.Windows.Forms.Message& m) [0x00000] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.Control+ControlWindowTarget.OnMessage (System.Windows.Forms.Message& m) [0x00000] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.Control+ControlNativeWindow.WndProc (System.Windows.Forms.Message& m) [0x0000b] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0  at System.Windows.Forms.NativeWindow.WndProc (System.IntPtr hWnd, System.Windows.Forms.Msg msg, System.IntPtr wParam, System.IntPtr lParam) [0x00085] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0Method '<Module>:__CxxQueryExceptionSize ()' in assembly '/home/skyshi/Desktop/Software/Bonsai/Packages/Bonsai.Spinnaker.0.7.1/build/net462/bin/x86/SpinnakerNET_v140.dll' contains native code that cannot be executed by Mono on this platform. The assembly was probably created using C++/CLI.Unhandled Exception:System.TypeInitializationException: The type initializer for '<Module>' threw an exception. ---> <CrtImplementationDetails>.ModuleLoadException: The C++ module failed to load. ---> System.MissingMethodException: Method contains unsupported native code assembly:<unknown assembly> type:<unknown type> member:(null)  at (wrapper managed-to-native) <Module>._getFiberPtrId()  at <Module>.<CrtImplementationDetails>.LanguageSupport._Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x0001c] in <0541963a254849e89638f912809c8605>:0  at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00028] in <0541963a254849e89638f912809c8605>:0   --- End of inner exception stack trace ---  at <Module>.<CrtImplementationDetails>.ThrowModuleLoadException (System.String errorMessage, System.Exception innerException) [0x00007] in <0541963a254849e89638f912809c8605>:0  at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00041] in <0541963a254849e89638f912809c8605>:0  at <Module>..cctor () [0x00008] in <0541963a254849e89638f912809c8605>:0   --- End of inner exception stack trace ---  at SpinnakerNET.ManagedSystem.!ManagedSystem () [0x00000] in <0541963a254849e89638f912809c8605>:0  at SpinnakerNET.ManagedSystem.Dispose (System.Boolean A_0) [0x0000c] in <0541963a254849e89638f912809c8605>:0  at SpinnakerNET.ManagedSystem.Finalize () [0x00000] in <0541963a254849e89638f912809c8605>:0[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeInitializationException: The type initializer for '<Module>' threw an exception. ---> <CrtImplementationDetails>.ModuleLoadException: The C++ module failed to load. ---> System.MissingMethodException: Method contains unsupported native code assembly:<unknown assembly> type:<unknown type> member:(null)  at (wrapper managed-to-native) <Module>._getFiberPtrId()  at <Module>.<CrtImplementationDetails>.LanguageSupport._Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x0001c] in <0541963a254849e89638f912809c8605>:0  at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00028] in <0541963a254849e89638f912809c8605>:0   --- End of inner exception stack trace ---  at <Module>.<CrtImplementationDetails>.ThrowModuleLoadException (System.String errorMessage, System.Exception innerException) [0x00007] in <0541963a254849e89638f912809c8605>:0  at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00041] in <0541963a254849e89638f912809c8605>:0  at <Module>..cctor () [0x00008] in <0541963a254849e89638f912809c8605>:0   --- End of inner exception stack trace ---  at SpinnakerNET.ManagedSystem.!ManagedSystem () [0x00000] in <0541963a254849e89638f912809c8605>:0  at SpinnakerNET.ManagedSystem.Dispose (System.Boolean A_0) [0x0000c] in <0541963a254849e89638f912809c8605>:0  at SpinnakerNET.ManagedSystem.Finalize () [0x00000] in <0541963a254849e89638f912809c8605>:0

--

**Reply by banchan86** [reply]
hmm I don't have experience with the spinnaker package on linux but you can download a 1.29.0.5 version for the SDK for linux on this page https://www.teledynevisionsolutions.com/products/spinnaker-sdk/?model=Spinnaker%20SDK&vertical=machine%20vision&segment=iis (requires a user account to be created). please report back if it works!

**Reply by skyshi0** [reply]
I downloaded the spinnaker 1.29.0.5 but it's designed for Ubuntu 18.04 and Ubuntu 16.04. The dependencies beyond Spinnaker 1.29.0.5 are unavailable now, I couldn't get the Spinnaker 1.29.0.5 running in my Ubuntu 22.04 Linux computer. 
![Screenshot from 2025-03-15 19-17-36](https://github.com/user-attachments/assets/864eb5f0-2c5a-4870-b044-aeaa6b9ecf58)
![Screenshot from 2025-03-15 19-41-41](https://github.com/user-attachments/assets/dcd7ad59-16ba-4d35-9f21-d59610af33ab)

**Reply by skyshi0** [reply]
@banchan86 Thank you for your reply, could you please upload the new spinnaker package here so I can try it later on my linux computer? [spinnaker package](https://github.com/bonsai-rx/spinnaker/pull/7)

**Reply by banchan86** [reply]
I think I have convinced @skyshi0 to switch back to Windows so this thread doesn't need a follow-up :smile:

**Comment by ncguilbeault** [comment]
Hi everyone, I've been working on a way to streamline installing/running Bonsai on Linux via Wine using the Nix package manager. The instructions are hopefully clear. It would be great if anyone wanted to give it a try and give feedback.

# Installing Bonsai on Linux using Nix
 
To install Bonsai, you must first install the nix package manager for your system. The instructions for this are here: https://nixos.org/download/. There are a few different options that you can select depending on your system, but the following steps should work in most cases, where you have sudo permissions and are on an x86_64 CPU architecture. Open up a terminal and run the following:
 
```bash
sh <(curl --proto '=https' --tlsv1.2 -L https://nixos.org/nix/install) --daemon
```
 
Once the script is downloaded, you will be prompted with a few questions by the nix installer. Select 'yes' for the options to run sudo and to add nixbld users. Once finished, open up a new terminal and confirm that nix is installed using `nix --version`. You should see something like: `nix (Nix) 2.34.7`. If it says `command not found`, then something may not have worked. If this doesn't work, then you need to troubleshoot the nix installation process before moving to the next step.
 
With nix now installed, run the following command:

```shell
nix --experimental-features "nix-command flakes" profile add github:ncguilbeault/bonsai-nix
```
 
When this runs for the first time, it will take a while. The reason is that it will download all of the dependencies (wine, winetricks, bonsai) and build them from source, making everything fully reproducible across systems. After the build step finishes, run the following from the terminal:

```bash
bonsai
```
 
When running this command for the first time, a setup script will run that will initialize the wine prefix, install DLLs into the prefix using winetricks, and then run the Bonsai installer. You will be prompted by the Bonsai installer to accept licenses and complete the installation. At the end, you will be prompted to either launch or exit. At this point, Bonsai should be fully installed and you can either launch it from the command line via the `bonsai` command, or launch it from the applications menu.

Let me know if you give it a try and if it works or doesn't work.

**Reply by chiyu1203** [reply]
Hi! I would love to give it a try. Just a quick question: why using  nix, instead of mono framework, to install Bonsai?

**Reply by ncguilbeault** [reply]
Nix is a package manager, so the goal here is just to package Bonsai together with its dependencies. It's similar to building a Debian package which you'd install with apt. The difference is that a Nix package is hermetic: every dependency is pinned inside it, all the way down to glibc, so it doesn't conflict with whatever you have on your system, or what your distro might ship. Once you have Nix itself, there isn't much else needed.

The other difference is that we use Wine as the translation layer rather than Mono. Mono was actually transferred to the WineHQ project a few years back, so most of the old Mono install/build instructions are stale. In effect, maintenance of Mono has moved to Wine, which now has a large, active developer base and is the defacto way to run Windows software on Linux (Valve's Proton, which powers the Steam Deck, is built on it).

In practice, you get most things working out of the box when using Wine. Bonsai's packages install identically and run pretty much the same as they do on Windows. Since Wine isn't doing emulation or virtualization, performance is also quite comparable to what you get on Windows. The main difficulty is that packages which depend on vendor SDKs or other Windows-specific drivers are either unsupported or require some workarounds.

**Reply by chiyu1203** [reply]
@ncguilbeault thank you very much for the detailed explanation. That makes a lot of sense! This discussion thread still keeps my question about some package compatibility to you few years ago. 
It was such a pain at that time. Thank you for putting this together.
I have tried the installation steps. It went smoothly and so did installing packages on Bonsai.
However, I bumped into problems with generating display (with BonVision package) and maybe connecting to the USB port (unable to acquire frames from some cheap webcams). Have you had these problems before?
<img width="854" height="585" alt="Screenshot from 2026-08-14 18-30-04" src="https://github.com/user-attachments/assets/319679c4-21c8-4753-817f-6f70407569c6" />
<img width="568" height="543" alt="Screenshot from 2026-08-14 18-28-53" src="https://github.com/user-attachments/assets/56e1cf7d-bc8b-4e6a-9ccf-ae7726f88a11" />
<img width="612" height="110" alt="Screenshot from 2026-08-14 18-39-19" src="https://github.com/user-attachments/assets/23ae5d49-cb22-45df-a8b8-674b9d8f3db1" />
